docs/reference/errors.md

```diff
@@ -0,0 +1,566 @@
+curly braces.
+
+Corrected message:
+
+<mf2-interactive>
+
+```mf2
+{|missing-brace|}
+```
+
+</mf2-interactive>
+
+## Data model errors
+
+Data model errors can be detected by the message formatter without
+knowing any runtime data. It's helpful to know this distinction because
+both syntax and data model errors are non-recoverable: the message
+formatter won't try to provide fallback output for messages that
+contain these errors.
+
+### Variant key mismatch
+
+JS name: `key-mismatch`
+
+A "Variant Key Mismatch" error occurs when the number of keys in a variant
+in a [matcher](/docs/reference/matcher/) differs from the number of selector variables.
+Remember that all the variants have to have the same number of keys,
+which has to be the same as the number of selector variables.
+
+<mf2-interactive>
+
+```mf2
+.local $x = {1 :number}
+.match $x
+one * {{Too many keys}}
+*   * {{Also too many keys}}
+```
+
+</mf2-interactive>
+
+#### Fixing the error
+
+Depending on what you intended, either change the number of selector variables
+to match the number of keys in the variants, or the other way around. Make
+sure that all variants have the same number of keys.
+
+Corrected message:
+
+<mf2-interactive>
+
+```mf2
+.local $x = {1 :number}
+.match $x
+one {{Correct number of keys}}
+*   {{Also correct number of keys}}
+```
+
+</mf2-interactive>
+
+### Missing fallback variant
+
+JS name: `missing-fallback`
+
+A "Missing Fallback Variant" error occurs when a [matcher](/docs/reference/matcher)
+isn't exhaustive. To be exhaustive, it must have one variant whose keys are all
+`*` (the wildcard key, which matches anything).
+
+<mf2-interactive>
+
+```mf2
+.local $x = {1 :number}
+.match $x
+one * {{x is singular}}
+```
+</mf2-interactive>
+
+#### Fixing the error
+
+Add a variant where all the keys are `*`. 
+
+<mf2-interactive>
+
+```mf2
+.local $x = {1 :number}
+.match $x
+one {{x is singular}}
+*   {{x is plural}}
+```
+</mf2-interactive>
+
+### Duplicate variant
+
+JS name: `duplicate-variant`
+
+A "Duplicate Variant" error occurs when more than one variant has the same
+list of keys. Key lists are ordered, so the key lists `one *` and `* one`
+would not be duplicates.
+
+<mf2-interactive>
+
+```mf2
+.local $x = {1 :number}
+.match $x
+one {{x is singular}}
+one {{x is singular.}}
+*   {{x is not singular}}
+```
+</mf2-interactive>
+
+This message is ambiguous because it's not clear which variant to choose
+if `$x` is singular.
+
+#### Fixing the error
+
+Remove the extra variant(s):
+
+<mf2-interactive>
+
+```mf2
+.local $x = {1 :number}
+.match $x
+one {{x is singular}}
+*   {{x is not singular}}
+```
+</mf2-interactive>
+
+
+### Missing selector annotation
+
+JS name: `missing-selector-annotation`
+
+<mf2-interactive>
+
+```mf2
+.match $x
+one {{x is singular}}
+*   {{x is plural}}
+```
+
+```json
+{ "x": 5 }
+```
+</mf2-interactive>
+
+This is one of the more complicated errors to understand. Every variable
```

Comment:
```suggestion
Every variable
```

I don't think docs should make the determination if someone finds something simple or difficult- they should generally be neutral to the reader